src/utils/GraphQLQueryBuilder.js

```diff
@@ -0,0 +1,155 @@
+const { AEM_GRAPHQL_TYPES } = require('./config')
+
+/**
+ *
+ * @private
+ * @param {object} obj - object representing query arguments
+ * @returns {string} - query args as a string
+ */
+function objToStringArgs (obj) {
+  let str = ''
+  for (const [key, value] of Object.entries(obj)) {
+    let val = typeof value === 'string' ? `"${value}"` : value
+    val = typeof value === 'object' ? `{ ${objToStringArgs(value)} }` : val
+    str += `${key}:${val}\n`
+  }
+  return str
+}
+
+/**
+ * Returns a Query for model by path
+ *
+ * @private
+ * @param {string} model - contentFragment Model Name
+ * @param {string} fields - The query string for item fields.
+ * @param {ModelByPathArgs} args - Query arguments
+ * @returns {QueryBuilderResult}
+ */
+const __modelByPath = (model, fields, args) => {
+  if (!args || !args._path) {
+    throw new Error('Missing required param "_path"')
+  }
+  const type = AEM_GRAPHQL_TYPES.BY_PATH
+  const query = `{
+    ${model}${type}(
+      ${objToStringArgs(args)}
+    ) {
+      item ${fields}
+    }
+  }`
+
+  return {
+    type,
+    query
+  }
+}
+
+/**
+ * Returns a Query for model list
+ *
+ * @private
+ * @param {string} model - contentFragment Model Name
+ * @param {string} fields - The query string for item fields.
+ * @param {ModelListArgs} [args={}] - Query arguments
+ * @returns {QueryBuilderResult}
+ */
+const __modelList = (model, fields, args = {}) => {
+  const type = AEM_GRAPHQL_TYPES.LIST
+  const query = `{
+    ${model}${type}(
+      ${objToStringArgs(args)}
+    ) {
+      items ${fields}
+    }
+  }`
+
+  return {
+    type,
+    query
+  }
+}
+
+/**
+ * Returns a Query for model list
+ *
+ * @private
+ * @param {string} model - contentFragment Model Name
+ * @param {string} fields - The query string for item fields.
+ * @param {ModelPaginatedArgs} [args={}] - Query arguments
+ * @returns {QueryBuilderResult}
+ */
+const __modelPaginated = (model, fields, args = {}) => {
+  const type = AEM_GRAPHQL_TYPES.PAGINATED
+  const query = `{
+    ${model}${type}(
+      ${objToStringArgs(args)}
+    ) {
+      pageInfo {
+        startCursor
+        endCursor
+        hasNextPage
+        hasPreviousPage
+      }
+      edges {
+        node ${fields}
+        cursor
+      }
+    }
+  }`
+
+  return {
+    type,
+    query
+  }
+}
+
+const getQueryType = (args = {}) => {
+  if (args._path) {
+    return AEM_GRAPHQL_TYPES.BY_PATH
+  }
+
+  if (args.useLimitOffset) {
+    return AEM_GRAPHQL_TYPES.LIST
+  }
+
+  return AEM_GRAPHQL_TYPES.PAGINATED
+}
+
+/**
+ * Builds a GraphQL query string for the given parameters.
+ *
+ * @param {string} model - contentFragment Model Name
+ * @param {ModelConfig} [config={}] - Pagination config
+ * @param {string} fields - The query string for item fields
+ * @param {ModelArgs} [args={}] - Query arguments
+ * @returns {QueryBuilderResult} - object with The GraphQL query string and type
+ */
+const graphQLQueryBuilder = (model, config = {}, fields, args = {}) => {
```

Comment:
I thought the same at first. Here are the reason why I selected this order
- model and config are both configuration related, while fields and args are strictly GraphQL query related.
- While `config` is optional, it's using default 10 items page size, and it would be good for devs to know that they are choosing default behaviour with providing null to config. 
- I would also expect that  in real usage `{ pageSize }` will always be part of config

But again this are minor reasons, if you still prefer the change I'll add it.

Reply:
Could it work if `fields` is an empty string? will it then return some default fields?
If that's possible, I would keep the signature as is and make fields optional as well.
If that's not, I would move fields to next to model. Because `fields` is mandatory anyway, I see little distinction between configuration and GraphQL.

Reply:
Order changed.